pkg/kubelet/kuberuntime/security_context.go

```diff
@@ -89,11 +89,14 @@ func verifyRunAsNonRoot(pod *v1.Pod, container *v1.Container, uid int64) error {
 		return nil
 	}
 
-	if uid == 0 {
+	switch {
+	case uid != nil && *uid == 0:
 		return fmt.Errorf("container has runAsNonRoot and image will run as root")
+	case uid == nil && len(username) > 0:
+		return fmt.Errorf("container has runAsNonRoot and image has non-numeric user (%s), cannot verify user is non-root", username)
+	default:
+		return nil
```

Comment:
two questions for @kubernetes/sig-node-pr-reviews 

1. what happens if there is no uid and no username? is that valid?
2. what happens if there is a non-zero uid and a username? does the numeric uid govern?

Reply:
I don't know the answers for sure but I'd expect the that behavior will mimic the Docker:

> what happens if there is no uid and no username? is that valid?

In this case container will be run with zero uid. And it isn't valid when `runAsNonRoot` is specified.

> what happens if there is a non-zero uid and a username? does the numeric uid govern?

AFIU uid is translated to `--user` options and overrides `USER` from Dockerfile:
```console
$ docker run --rm -it --entrypoint /usr/bin/id phpcoder/mystamps 
uid=1000(mystamps) gid=1000(mystamps) groups=1000(mystamps)
$ docker run --rm -u 2000 -it --entrypoint /usr/bin/id phpcoder/mystamps 
uid=2000 gid=0(root)
```

Reply:
>> what happens if there is no uid and no username? is that valid?
>
> In this case container will be run with zero uid. And it isn't valid when runAsNonRoot is specified.

I'd expect that to be invalid as well, so we'd need this case:
```
case uid == nil && len(username) == 0:
  return fmt.Errorf("container has runAsNonRoot and no uid or username, cannot verify user is non-root", username)
```

@sjenning, does that sound right to you?

Reply:
old code behavior was
1. no validation; assumed valid
2. username is ignored if uid != nil

looks to me that the old behaviour is maintained in this refactor.

Reply:
> I'd expect that to be invalid as well, so we'd need this case:

I can add it but it should never happens because `GetImageUser` that we use for getting uid/username returns zero uid when uid/username couldn't be determined:

https://github.com/kubernetes/kubernetes/blob/a0ed61671cb86f26dbce128be857a95a3d33b5ec/pkg/kubelet/kuberuntime/helpers.go#L130-L150

Reply:
ok, this is fine as-is then